lib/fast_jsonapi/relationship.rb

```diff
@@ -136,13 +158,39 @@ def id_hash(id, record_type, default_return = false)
     end
 
     def fetch_id(record, params)
-      if object_block.present?
-        object = FastJsonapi.call_proc(object_block, record, params)
-        return object.map { |item| item.public_send(id_method_name) } if object.respond_to? :map
+      if object_block.nil?
+        #
+        # TODO: This line assumes the `id_method_name` is a foreign key defined on the parent record.
+        #       The `id_method_name` option should be replaced with `foreign_key` to make the intent
+        #       explicit.
+        #
+        foreign_key = id_method_name
+        foreign_key ||= (has_many? ? "#{@name.to_s.singularize}_ids" : "#{@name}_id")
+
+        return record.public_send(foreign_key) if record.respond_to?(foreign_key)
+      end
+
+      return unless object = fetch_associated_object(record, params)
+
+      #
+      # TODO: The serializer for a relationship should always be the single source of truth for determining the ID of
+      #       a related record. This code path assumes that the `id_method_name` should be called on the child record.
+      #       This code path should be removed when `id_method_name` is replaced with `foreign_key`
+      #
+      if id_method_name
+        if has_many? && object.first.respond_to?(id_method_name)
```

Comment:
Thanks for taking a stab at this, I think it's really close!

I'm a bit concerned about calling `first` here.  If it's an Array, no big deal.  If it's an ActiveRecord relationship, and the relationship has not been loaded yet, then `first` will call its own `SELECT ... LIMIT 1` query.  Then the full query of all records would follow when `map` is called.  So you'd get two queries here, which we should avoid.

Also, what happens if the developer intends for `id_method_name` to be called on the child, but the relationship has no associated records?  It will fall through to attempting the `id_method_name` on the parent, which may result in incorrect output.

I think we can solve for both this way:

```ruby
if has_many?
  if object.length > 0 && object.first.respond_to?(id_method_name)
    return object.map { |item| item.public_send(id_method_name) }
  else
    return []
  end
...
end
```

Also, what should happen if `id_method_name` is specified but neither the parent nor child respond to it?  I tend to think it should throw an exception as opposed to falling through to the next attempts.  If the developer specifies it, it should tell the developer if it's wrong.